### ISSUE TITLE
Show count and cursor position in the header

### ISSUE THREAD

**Issue by meirotstein** [josdejong/jsoneditor#443]
Hello,

It would be useful to add an option to show in the header the characters count and the cursor position (line/column) such as in other editors.

If you find it useful as well perhaps I'll find the time to contribute it myself.

Thanks!
Meir.

**Comment by josdejong** [comment]
Thanks for your suggestion Meir.

Most text editors and IDE's show the cursor location rows/cols. I don't think that is really helpful for JSONEditor since both object properties and values are typically short words or sentences. What may be more interesting is displaying the current path on top, right below the main menu. What do you think?

**Comment by meirotstein** [comment]
This is very interesting suggestion.. however I'm afraid it may have some performance impact as it require to calculate the cursor location on every move, don't you think?

**Comment by josdejong** [comment]
I don't think performance will be an issue here, only on `focus` and `blur` of input fields we have to re-render the current path.

**Comment by meirotstein** [comment]
My concern is mainly in regard to the code editor (I'm building a tool where it is sometimes important to know the characters count) - in that case every cursor move will trigger a re-render. 

perhaps it will be best to add a cursor location and count to the code editor and a path to the form editor..

**Comment by josdejong** [comment]
Ah, good point. For the code editor we could debounce changes whilst typing, and as soon as the cursor hasn't changed for say 300 ms, update the path.

**Comment by meirotstein** [comment]
Got it.
(IMHO for code editor row/col is better, however - you're the boss :) )

I will start work on it and push a PR once it is done.
Thanks!

**Comment by josdejong** [comment]
> for code editor row/col is better

yeah. Maybe we should have a row/col for code  and text mode, and a path for tree mode?

**Comment by meirotstein** [comment]
Yes. for both tree & form modes.

**Comment by meirotstein** [comment]
Hi @josdejong ,

Better late than never :)

I have created a PR which contains the following:

**For text and code** - when selecting text it shows the count of the selected characters on the menu 
**For code** - the line and column of the cursor were added to the menu

I have decided not to add line and column for text editor as it uses textarea and the re-calculation for each key or mouse change seems to me quite expansive, I think it is also OK the way it is..

These additions are applied by default with no option to control it, please let me know what you think about that.

**For tree, view and form** - a new option is available - treePath. default is false.
when setting it to true it adds a new control to the menu right below the current one - it shows a tree path control that does the following:

- Visualize the location of the current selection in the tree structure 

- Allows to focus other nodes by selecting it on the tree path

- Allows breadcrumbs navigation

Please review and let me know what you think

Thanks,
Meir.

**Comment by josdejong** [comment]
cool, thanks! I will review your PR soon

**Comment by meirotstein** [comment]
A CodePen for demonstration:
https://codepen.io/meirotstein/pen/yomZXq

**Comment by josdejong** [comment]
This works really cool Meir  :sunglasses:. It's a way nicer addition than I had expected. The dropdowns in the navigation work very nice. 

Some feedback:
1. I'm not sure if the name of the option `treePath` is very clear, how about `navigationBar` or something like that?
2. There is currently a CSS issue: the contents of the tree itself overflow the JSONEditor boundaries due to the navigationBar. 
3. I think we need to do some refinement in the styling of the navigation bar. Maybe a light gray background and dark gray text color looks a bit more clean? And maybe instead of the double arrow (which is a bit noisy) use a single arrow or a filled right pointing triangle?
4. As for the cursor position in code/text mode: I've never seen this on top, it's typically displayed in a bar at the bottom. Whilst the current solution saves space, I think it will look more neat to display it in a new bar at the bottom (light gray background, darkgray font). What do you think?
5. Ideally the text mode should have row/col information too, like the code mode.
6. I think we should make displaying row/col in code mode an option too (`statusBar: true` or something like that), what do you think?

**Comment by meirotstein** [comment]
Hello,

Thanks for the review! - as for you comments:

1. `navigationBar` does sounds better - will change it.
2. RIght... will be fixed.
3. OK - I will work on it, if you can provide any mock or example of a similar tool it will be great. as for the double arrow - I tried first a single arrow however it was too small for a clickable element. a triangle sounds like a good idea - will try it.
4. I agree. a bottom bar is a good idea. actually I debated it myself but decided to integrate the existing layout, however the bottom bar approach is more clean - will do that.
5. I agree. however this is a non trivial addition for textarea - I suggest that we'll split it to a different PR that I will add later, what do you think?
6. Sure - in my opinion it should be true by default - what do you think?

**Comment by josdejong** [comment]
3. For example WebStorm has such a navigation bar. Windows Explorer I think but I'm not running windows right now so I can't test. A small arrow or triangle does not need to be a problem if you create a button with enough padding. For example a button can be 24x24px and have a small triangle of 8x8px in the middle, like the expand/collaps triangles in tree mode. Here a screenshot of the navigation bar of WebStorm:
    
    ![image](https://user-images.githubusercontent.com/568626/30397943-cf9e3e6c-98ce-11e7-9628-5186cf873f61.png)

4. awesome
5. Yes it's ok with me to do implement this for text mode in a separate PR, this one is already quite large ;)
6. Yes, I think `navigationBar` and `statusBar` should be `true` by default.

**Comment by meirotstein** [comment]
Hi @josdejong ,

I have made some changes, as we discussed 

You can can see it on the CodePen:
https://codepen.io/meirotstein/pen/yomZXq

Waiting for your review,
thanks!

**Comment by josdejong** [comment]
Looking at your CodePen... that looks great! Thanks

Maybe some slight color tweaking: can you try font color `gray` (`#808080`) and no blue borders for the navigation bar and status bar?

I will check your code changes in more detail soon.

**Comment by meirotstein** [comment]
As for your suggestion - if there's no border it is also recommended to change the background to a slightly darker gray - say #dcdcdc - this way the statusBar is a bit different then the left bar (row numbers), something like that:

**navigationBar**:

![image](https://user-images.githubusercontent.com/2209844/30613928-763fb940-9d92-11e7-9c0a-c35daeb924dc.png)

**statusBar**:

![image](https://user-images.githubusercontent.com/2209844/30613956-93560002-9d92-11e7-913d-28140610e0a3.png)

What do you think?

**Comment by josdejong** [comment]
Yes looks good :+1:

**Comment by fiddej** [comment]
Is there a way to get the cursor position's line number as a variable? 

For example like: 
```
    var container = document.getElementById("myDiv");
    var options = { mode : "code" };
    var editor= new JSONEditor(container , options);
    var lineNumber = editor.getCursorLine();
```

**Comment by meirotstein** [comment]
Hi,

For code editor the cursor position can be given quite easily - as for the char (do u mean the char at the cursor location? the selection?) - this requires a bit more effort.
For textarea it will require additional effort that is not part of this change.

Perhaps you can elaborate a little more in regard for your requirement.

As a workaround you can get the cursor position from the aceEditor instance that it used in code mode:
`editor.aceEditor.getCursorPosition()`

anyhow AFAIS this is not part of the official jsoneditor API..

**Comment by fiddej** [comment]
Thank you meirotstein,

I solved it with editing the jsoneditor.js slightly. I simply saved the 

`var curserPos = this.aceEditor.getCursorPosition();` (misspelled curserPos?) 

in a new variable on the JSONEditor object too. Then could access the row and col. By char I meant the total number of chars prior the cursor, ignoring row breaks. But I'm good for now. Thanks.

**Comment by meirotstein** [comment]
https://mightyredpen.wordpress.com/2007/12/05/curser-versus-cursor/

:)

**Comment by josdejong** [comment]
It's an interesting idea to extend the API with functions to get the current cursor position (and I think we would need a callback triggered when the position changes too), but I suggest we first finish this PR first and implement that as a second step.

**Comment by josdejong** [comment]
@meirotstein as for your latest tweaks: I think it looks great now, thanks :). I think it's about ready to merge.

I will do a little bit of tweaking of the font-size afterwards, the editor contains some different sizes (also already in the mode menu, that needs a bit of attention too)

There is one last detail: when the path is so long that it doesn't fit on one line, it wraps to the second line but that doesn't look good since the height of the navigationBar is fixed. I don't know an easy solution here, and I think this is an edge case which we shouldn't over engineer. This sort of things can be solved with flex box, but that is not supported by IE10 and it would be a bit of a pity to drop support of IE10 just for this edge case. Maybe we can add a few lines of JavaScript after updating the path to measure the actual height and adjust it's fixed height and the top margin/padding of the contents or something? Do you have a creative idea?

**Comment by meirotstein** [comment]
Hi @josdejong ,

Thanks for the review - I'm glad that you like it!

As for the long path - I have changed the behaviour few comments ago from the same reason - currently the path is not wrapped to a second lline (and third and so on), instead it scrolls the path to the selection (see [CodePen](https://codepen.io/meirotstein/pen/yomZXq)) - you'll alway see the end of the path.

For instance:
path: `a>b>c>d>e>f>g`
navigationBar space fits to 4 path elements only
You'll get (visible parts are between the [] signs): `a>b>c>[d>e>f>g]`

perhaps adding an ellipsis to the beginning of the path in such case will give a better experience (it not seems critical to me) 

p.s. - I hates IE! :)

**Comment by josdejong** [comment]
ok, let's keep that open for further refinement.

I've merged your work now (#457), thanks again Meir! Coming days I will do a bit of refinement in the UI overall an some more testing (all examples, different browsers) an publish it.

I've opened two followup issues of ideas that popped up during our discussion here:
- #470 API for selection
- #471 caret position in status bar when in mode `text`

**Comment by josdejong** [comment]
@meirotstein I thought that I had already released the next version with navigation bar and status bar but that wasn't the case :flushed: sorry.

I've refined the CSS a little bit (looks again a bit more like one of your earlier versions, this works now because I darkened the color of selected text a bit), I hope you like it this way. You can see the result on http://jsoneditoronline.org/ Thanks again!

**Comment by meirotstein** [comment]
Thanks - looks good!
